app/lib/dartdoc/backend.dart

```diff
@@ -320,3 +320,58 @@ class DartdocBackend {
     await deletePool.close();
   }
 }
+
+class ArchiveTracker {
+  final _pvMap = <String, _TrackedEntry>{};
+  final _sampleLines = <String>[];
+
+  ArchiveTracker() {
+    _logger.info('Tracking: init.');
+    new Timer.periodic(const Duration(minutes: 10), (_) => _cleanup());
+    new Timer.periodic(const Duration(minutes: 45), (_) => _logSample());
+  }
+
+  List<String> get samples => _sampleLines;
+
+  void track(DartdocEntry entry) {
+    final key = '${entry.packageName}/${entry.packageVersion}';
+    final tracked = _pvMap.putIfAbsent(key, () => new _TrackedEntry());
+    tracked.archiveSize = entry.archiveSize ?? 0;
+    tracked.totalSize = entry.totalSize ?? 0;
+    tracked.lastAccess = new DateTime.now();
+  }
+
+  void _cleanup() {
+    final now = new DateTime.now();
+    _pvMap.removeWhere((key, value) {
+      final age = value.lastAccess.difference(now).abs();
+      return age.inMinutes > 90;
+    });
```

Comment:
This is a linear scan and doesn't scale well if there are a lot of entries. You could use a Queue that lets you efficiently dequeue the oldest entries. However, I don't think Dart's Queue class lets you efficiently find and remove an object from the middle (so it can be inserted at the end). Perhaps there's a priority queue class out there that could be used. But if the cache isn't too big, this can be fine.

What is being measured here? This cleanup seems detached from other stuff, like some cache whose size is being measured. Is it an attempt at a moving window statistic? You could have buckets with per-minute or per-15-minutes statistics that you just add to (though you'll probably want to keep track of which entries have been added to those buckets).

Reply:
The thought process behind the cleanup is the following: when we have an access to a `DartdocEntry`, it would mean that we either download its related package archive to local disk, or we have it already there. After 90 minutes of inactivity we would delete it from the local disk. The cleanup is done in 10 minutes intervals, making sure we don't stretch too much over the 90 minutes, but the sampling of the required space is not important to be precise.